# Generation G - S1E1 - 18 lines app

This notebook is the companion of posts about Generative AI.

This episode shows how to create a very simple app in 18 lines, and the implications of running this app, especially from the FinOps point of view.

## LangChain

LangChain is a new open source framework that allows AI developers to combine Large Language Models (LLMs) with external data.

LangChain resources
> - Landpage: https://readthedocs.org/projects/langchain/db2d
> - Comonents: https://docs.langchain.com/docs/category/components
> - git: https://github.com/hwchase17/langchain.git
> - API Reference: https://api.python.langchain.com/en/latest/

## The 18 lines streamlit app

Let's start with the [basic LLM app in 18 lines of code from the 
streamlit documentation](https://blog.streamlit.io/langchain-tutorial-1-build-an-llm-powered-app-in-18-lines-of-code/)

Here is the full app code. It is really minimalistic, a text box for the question, the response is displayed right below.

```python
import streamlit as st
from langchain.llms import OpenAI

st.title('🦜🔗 Quickstart App')

openai_api_key = st.sidebar.text_input('OpenAI API Key')

def generate_response(input_text):
  llm = OpenAI(temperature=0.7, openai_api_key=openai_api_key)
  st.info(llm(input_text))

with st.form('my_form'):
  text = st.text_area('Enter text:', 'What are the three key pieces of advice for learning how to code?')
  submitted = st.form_submit_button('Submit')
  if not openai_api_key.startswith('sk-'):
    st.warning('Please enter your OpenAI API key!', icon='⚠')
  if submitted and openai_api_key.startswith('sk-'):
    generate_response(text)
```

## The app's interaction with OpenAI APIs

The interaction with the model is located in the function __generate_response__. Does not seem to hard to workaround.

```python
def generate_response(input_text):
    llm = OpenAI(temperature=0.7, openai_api_key=openai_api_key)
    st.info(llm(input_text))
```

## notes for conclusion 


# Material

## Initializations

### Update environment

In [9]:
!apt-get update && apt-get install -y build-essential 1>/dev/null

Hit:1 http://deb.debian.org/debian bullseye InRelease
Hit:2 http://deb.debian.org/debian bullseye-updates InRelease
Hit:3 http://security.debian.org/debian-security bullseye-security InRelease
Reading package lists... Done


In [10]:
!apt-get update && apt-get install -y jq 1>/dev/null

Hit:1 http://deb.debian.org/debian bullseye InRelease
Hit:2 http://deb.debian.org/debian bullseye-updates InRelease
Hit:3 http://security.debian.org/debian-security bullseye-security InRelease
Reading package lists... Done


In [11]:
!pip install --upgrade pip  1>/dev/null

## Requirements

In [12]:
!pip install langchain==0.0.230 1>/dev/null

In [13]:
!pip install openai==0.27.8 1>/dev/null

In [40]:
!pip install tiktoken==0.4.0 1>/dev/null

## Secrets and credentials

In [14]:
%%bash --out secrets 
# using AWS's Secret Manager to store keys
# garb the keys and store it into a Pytthon variable
export RESPONSE=$(aws secretsmanager get-secret-value --secret-id 'salvia/labbench/tests' )
export SECRETS=$( echo $RESPONSE | jq '.SecretString | fromjson')

echo $SECRETS

In [15]:
import os

os.environ["OPENAI_API_KEY"] = eval(secrets)["OPENAI_API_KEY"]


## Code session

change in generate_response

```python
def generate_response(input_text):
    llm = get_llm_model()
    st.info(llm(input_text))
```

Tests of the llm in isolation

Resources
> - https://api.python.langchain.com/en/latest/llms/langchain.llms.openai.OpenAI.html

In [17]:
%%time
import os
from langchain.llms import OpenAI
from langchain.llms.fake import FakeListLLM

openai_api_key = os.environ["OPENAI_API_KEY"] 

def get_llm_model():
    llm = OpenAI(temperature=0.7, openai_api_key=openai_api_key)
    return llm

llm = get_llm_model()

CPU times: user 10 ms, sys: 4.7 ms, total: 14.7 ms
Wall time: 42.8 ms


In [34]:
llm_info = llm.json
#print(llm_info)  # warning: disclose the apo key

<bound method BaseModel.json of OpenAI(cache=None, verbose=False, callbacks=None, callback_manager=None, tags=None, metadata=None, client=<class 'openai.api_resources.completion.Completion'>, model_name='text-davinci-003', temperature=0.7, max_tokens=256, top_p=1, frequency_penalty=0, presence_penalty=0, n=1, best_of=1, model_kwargs={}, openai_api_key='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', openai_api_base='', openai_organization='', openai_proxy='', batch_size=20, request_timeout=None, logit_bias={}, max_retries=6, streaming=False, allowed_special=set(), disallowed_special='all', tiktoken_model_name=None)>


In [37]:
print(f"max tokens response {llm.max_tokens}")

max tokens response 256


In [38]:
print(f"max tokens total {llm.max_context_size}")

max tokens total 4097


In [41]:
%%time
query = "What is the distance to the Moon?"
response = llm(query)
print(f"query tolens {llm.get_num_tokens(query)} - result tokens {llm.get_num_tokens(response)}")
print(response)

query tolens 8 - result tokens 21


The average distance from Earth to the Moon is 238,855 miles (384,400 kilometers).
CPU times: user 254 ms, sys: 53.7 ms, total: 307 ms
Wall time: 2.37 s


In [ ]:
get_num_tokens(text: str) → int¶


In [42]:
%%time
query = "Who is the White Rabbit?"
response = llm(query)
print(f"query tolens {llm.get_num_tokens(query)} - result tokens {llm.get_num_tokens(response)}")
print(response)

query tolens 6 - result tokens 78


The White Rabbit is a fictional character in Lewis Carroll's 1865 novel Alice's Adventures in Wonderland. The White Rabbit is constantly late and in a state of frenzied panic. He wears a waistcoat and carries a pocket watch. He is very worried about being late, and is often seen muttering to himself, "Oh dear! Oh dear! I shall be too late!"
CPU times: user 6.43 ms, sys: 387 µs, total: 6.82 ms
Wall time: 1.83 s



To the question "What is the distance to the Moon?", we get the right answer:
> "The average distance from Earth to the Moon is 238,855 miles (384,400 kilometers)".

And to the question  "Who is the White Rabbit?", we get a typical ChatGPT answer:
> The White Rabbit is a fictional character from the book Alice's Adventures in Wonderland by Lewis Carroll. He appears at the very beginning of the story, in which Alice follows him down a rabbit hole, and is noted for his continual use of the phrase "Oh dear! Oh dear! I shall be too late!" He is often referred to as the March Hare, due to his association with the March Hare in the book's sequel, Through the Looking-Glass.

## OpenAI usage 

https://platform.openai.com/account/usage

## AWS usage

https://platform.openai.com/account/usage